Docs/source/usage/workflows/extend.rst

```diff
@@ -0,0 +1,229 @@
+Important APIs include:
+
+* `amr.ParallelDescriptor <https://pyamrex.readthedocs.io/en/latest/usage/api.html#amrex.space3d.ParallelDescriptor.IOProcessor>`__: MPI-parallel rank information
+* `amr.MultiFab <https://pyamrex.readthedocs.io/en/latest/usage/api.html#amrex.space3d.MultiFab>`__: MPI-parallel field data
+* `amr.ParticleContainer_* <https://pyamrex.readthedocs.io/en/latest/usage/api.html#amrex.space3d.ParticleContainer_1_1_2_1_default>`: MPI-parallel particle data for a particle species
+
+
+Data Access
+-----------
+
+While the simulation is running, callbacks can read and write access the WarpX simulation data *in situ*.
+
+An important object in the ``pywarpx.picmi`` module for data access is ``Simulation.extension.warpx``, which is available only during the simulation run.
+This object is the Python equivalent to the C++ ``WarpX`` simulation class.
+
+.. py:class:: WarpX
+
+   .. py:method:: getistep(lev: int)
+
+      Get the current step on mesh-refinement level ``lev``.
+
+   .. py:method:: gett_new(lev: int)
+
+      Get the current physical time on mesh-refinement level ``lev``.
+
+   .. py:method:: getdt(lev: int)
+
+      Get the current physical time step size on mesh-refinement level ``lev``.
+
+   .. py:method:: multifab(multifab_name: str)
+
+      Return MultiFabs by name, e.g., ``"Efield_aux[x][l=0]"``, ``"Efield_cp[x][l=0]"``, ...
+
+      The physical fields in WarpX have the following naming:
+
+      - ``_fp`` are the "fine" patches, the regular resolution of a current mesh-refinement level
+      - ``_aux`` are temporary (auxiliar) patches at the same resolution as ``_fp``.
+        They usually include contributions from other levels and can be interpolated for gather routines of particles.
+      - ``_cp`` are "coarse" patches, at the same resolution (but not necessary values) as the ``_fp`` of ``level - 1``
+        (only for level 1 and higher).
+
+   .. py:method:: multi_particle_container
+
+   .. py:method:: get_particle_boundary_buffer
+
+   .. py:method:: set_potential_on_eb(potential: str)
+
+      The embedded boundary (EB) conditions can be modified when using the electrostatic solver.
+      This set the EB potential string and updates the function parser.
+
+   .. py:method:: evolve(numsteps=-1)
+
+      Evolve the simulation the specified number of steps.
+
+   .. autofunction:: pywarpx.picmi.Simulation.extension.finalize
+
+.. py::def:: pywarpx.picmi.Simulation.extension.get_instance
+
+   Return a reference to the WarpX object.
+
+
+The :py:class:`WarpX` also provides read and write access to field ``MultiFab`` and ``ParticleContainer`` data, shown in the following examples.
+
+Fields
+^^^^^^
+
+This example accesses the :math:`E_x(x,y,z)` field at level 0 after every time step and calculate the largest value in it.
+
+.. code-block:: python3
+
+   # Preparation: set up the simulation
+   #   sim = picmi.Simulation(...)
+   #   ...
+
+
+   @callfromafterstep
+   def sum_rho_electrons():
```

Comment:
This function name should probably be changed to something more descriptive of the actual function, perhaps `get_max_E_x()`